File: cloudflare-workers/import-notes.js
#!/usr/bin/env node
/**
 * 批量导入笔记到Cloudflare Vectorize
 * 
 * 使用方法：
 * 1. 先在本地启动wrangler dev
 * 2. 运行: node import-notes.js
 */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import sqlite3 from 'better-sqlite3';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// 本地数据库路径
const DB_PATH = path.join(__dirname, '..', 'data', 'xhs_notes.db');

// API地址（Cloudflare生产环境）
const API_BASE = 'https://xhs-rag.ok2442504.workers.dev';

/**
 * 从本地SQLite加载笔记
 */
function loadNotesFromDB() {
    if (!fs.existsSync(DB_PATH)) {
        console.error('数据库文件不存在:', DB_PATH);
        console.log('请先在本地RAG系统中采集笔记');
        return [];
    }
    
    const db = sqlite3(DB_PATH);
    
    const notes = db.prepare(`
        SELECT 
            note_id, 
            title, 
            content, 
            url, 
            user_id as author,
            nickname,
            liked_count,
            collected_count,
            publish_time,
            created_at
        FROM notes
    `).all();
    
    db.close();
    
    console.log(`从数据库加载了 ${notes.length} 篇笔记`);
    return notes;
}

/**
 * 批量添加到API
 */
async function batchAddNotes(notes, batchSize = 10) {
    const results = {
        success: 0,
        failed: 0,
        errors: []
    };
    
    for (let i = 0; i < notes.length; i += batchSize) {
        const batch = notes.slice(i, i + batchSize);
        const batchNumber = Math.floor(i / batchSize) + 1;
        const totalBatches = Math.ceil(notes.length / batchSize);
        
        console.log(`\n处理批次 ${batchNumber}/${totalBatches} (${batch.length} 条笔记)`);
        
        try {
            const response = await fetch(`${API_BASE}/api/notes/batch`, {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify({ notes: batch })
            });
            
            const result = await response.json();
            
            if (result.successful) {
                results.success += result.successful;
                console.log(`  ✅ 成功添加 ${result.successful}/${batch.length} 条`);
            }
            
            if (result.results) {
                result.results.forEach(r => {
                    if (!r.success) {
                        results.failed++;
                        results.errors.push({
                            note_id: r.note_id,
                            error: r.error
                        });
                    }
                });
            }
            
        } catch (error) {
            console.error(`  ❌ 批次 ${batchNumber} 请求失败:`, error.message);
            results.failed += batch.length;
            results.errors.push({
                batch: batchNumber,
                error: error.message
            });
        }
        
        // 避免请求过快
        await new Promise(resolve => setTimeout(resolve, 1000));
    }
    
    return results;
}

/**
 * 主函数
 */
async function main() {
    console.log('================================');
    console.log('笔记批量导入工具');
    console.log('================================\n');
    
    // 检查API连接
    console.log('检查API连接...');
    try {
        const healthRes = await fetch(`${API_BASE}/health`);
        if (healthRes.ok) {
            console.log('✅ API已连接\n');
        } else {
            throw new Error('API返回错误状态');
        }
    } catch (error) {
        console.error('❌ API连接失败:', error.message);
        console.log('\n请确保：');
        console.log('1. 已运行 npx wrangler dev');
        console.log('2. Wrangler服务正在 http://localhost:8787 运行\n');
        process.exit(1);
    }
    
    // 加载笔记
    const notes = loadNotesFromDB();
    
    if (notes.length === 0) {
        console.log('没有找到可导入的笔记');
        process.exit(0);
    }
    
    // 确认导入
    console.log('\n即将导入的笔记预览：');
    notes.slice(0, 5).forEach((note, i) => {
        console.log(`  ${i + 1}. ${note.title.substring(0, 50)}...`);
    });
    if (notes.length > 5) {
        console.log(`  ... 还有 ${notes.length - 5} 条`);
    }
    
    // 开始导入
    console.log('\n开始导入...\n');
    const results = await batchAddNotes(notes);
    
    // 显示结果
    console.log('\n================================');
    console.log('导入完成！');
    console.log(`✅ 成功: ${results.success}`);
    console.log(`❌ 失败: ${results.failed}`);
    
    if (results.errors.length > 0) {
        console.log('\n错误详情：');
        results.errors.slice(0, 10).forEach(err => {
            console.log(`  - ${err.note_id || '未知'}: ${err.error}`);
        });
    }
    console.log('================================');
}

// 运行
main().catch(console.error);

File: cloudflare-workers/import_notes.py
#!/usr/bin/env python
"""
批量导入笔记到Cloudflare Vectorize
在本地使用sentence-transformers生成中文嵌入
"""

import os
import sys
import json
import time
import sqlite3

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

try:
    import requests
except ImportError:
    print("请先安装requests: pip install requests")
    sys.exit(1)

try:
    from sentence_transformers import SentenceTransformer
    MODEL = SentenceTransformer('paraphrase-multilingual-MiniLM-L12-v2')
    print('[OK] 加载语义嵌入模型: paraphrase-multilingual-MiniLM-L12-v2')
except ImportError:
    print("请先安装sentence-transformers: pip install sentence-transformers")
    sys.exit(1)

DB_PATH = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), 'data', 'xhs_notes.db')
API_BASE = 'https://xhs-rag.ok2442504.workers.dev'


def generate_embedding(text):
    return MODEL.encode(text).tolist()


def load_notes_from_db():
    if not os.path.exists(DB_PATH):
        print(f'数据库文件不存在: {DB_PATH}')
        return []
    
    conn = sqlite3.connect(DB_PATH)
    conn.row_factory = sqlite3.Row
    
    cursor = conn.execute('''
        SELECT note_id, title, content, url, user_id as author, nickname, 
               liked_count, collected_count, publish_time, created_at
        FROM notes
    ''')
    
    notes = []
    for row in cursor.fetchall():
        notes.append({
            'note_id': row['note_id'],
            'title': row['title'] or '',
            'content': row['content'] or '',
            'url': row['url'] or '',
            'author': row['author'] or '',
        })
    
    conn.close()
    print(f'从数据库加载了 {len(notes)} 篇笔记')
    return notes


def batch_add_notes_with_embeddings(notes, batch_size=5):
    success = 0
    failed = 0
    errors = []
    
    for i in range(0, len(notes), batch_size):
        batch = notes[i:i+batch_size]
        batch_num = i // batch_size + 1
        total_batches = (len(notes) + batch_size - 1) // batch_size
        
        print(f'\n处理批次 {batch_num}/{total_batches} ({len(batch)} 条笔记)')
        
        try:
            notes_with_embeddings = []
            for note in batch:
                content = note["content"] or ''
                truncated_content = content[:2000]
                text_to_embed = f'{note["title"]}\n\n{truncated_content}'
                embedding = generate_embedding(text_to_embed)
                notes_with_embeddings.append({
                    'note_id': note['note_id'],
                    'title': note['title'],
                    'content': truncated_content,
                    'url': note['url'],
                    'author': note['author'],
                    'embedding': embedding,
                })
            
            response = requests.post(
                f'{API_BASE}/api/notes/batch-with-embedding',
                json={'notes': notes_with_embeddings},
                headers={'Content-Type': 'application/json'},
                timeout=30
            )
            
            if response.status_code == 200:
                result = response.json()
                if result.get('successful'):
                    success += result['successful']
                    print(f'  ✅ 成功添加 {result["successful"]}/{len(batch)} 条')
                if result.get('results'):
                    for r in result['results']:
                        if not r['success']:
                            failed += 1
                            errors.append({'note_id': r['note_id'], 'error': r['error']})
            else:
                print(f'  ❌ 请求失败: {response.status_code} - {response.text}')
                failed += len(batch)
                
        except Exception as e:
            print(f'  ❌ 批次 {batch_num} 请求失败: {e}')
            failed += len(batch)
        
        time.sleep(0.5)
    
    return {'success': success, 'failed': failed, 'errors': errors}


def main():
    print('================================')
    print('笔记批量导入工具（本地中文嵌入）')
    print('================================\n')
    
    print('检查API连接...')
    try:
        response = requests.get(f'{API_BASE}/health', timeout=10)
        if response.status_code == 200:
            print('✅ API已连接\n')
        else:
            raise Exception(f'API返回错误状态: {response.status_code}')
    except Exception as e:
        print(f'❌ API连接失败: {e}')
        print('\n请确保：')
        print('1. Cloudflare Worker已部署')
        print(f'2. API地址正确: {API_BASE}\n')
        sys.exit(1)
    
    notes = load_notes_from_db()
    
    if not notes:
        print('没有找到可导入的笔记')
        sys.exit(0)
    
    print('\n即将导入的笔记预览：')
    for i, note in enumerate(notes[:5], 1):
        print(f'  {i}. {note["title"][:50]}...')
    if len(notes) > 5:
        print(f'  ... 还有 {len(notes) - 5} 条')
    
    print('\n开始导入（本地生成中文嵌入）...\n')
    results = batch_add_notes_with_embeddings(notes)
    
    print('\n================================')
    print('导入完成！')
    print(f'✅ 成功: {results["success"]}')
    print(f'❌ 失败: {results["failed"]}')
    
    if results['errors']:
        print('\n错误详情：')
        for err in results['errors'][:10]:
            print(f'  - {err["note_id"]}: {err["error"]}')
    print('================================')


if __name__ == '__main__':
    main()
